src/builder/commands/copy.rs

```diff
@@ -94,16 +143,38 @@ impl BuildStep for Copy {
     {
         let ref src = self.source;
         if src.starts_with("/work") {
-            let filter = try!(Filter::new(&self.ignore_regex, &self.include_regex));
-            try!(hash_path(hash, src, &filter, self.owner_uid, self.owner_gid));
+            let filter = try!(Filter::new(
+                &self.ignore_regex, &self.include_regex));
+            try!(hash_path(hash, src, &filter, |h, p, st| {
+                h.field("filename", p.as_os_str().as_bytes());
+                if let Some(mode) = self.calc_mode(st) {
+                    h.text("mode", mode);
+                }
+                h.text("uid", self.owner_uid.unwrap_or(st.uid()));
+                h.text("gid", self.owner_gid.unwrap_or(st.gid()));
+                try!(hash_file_content(h, p, st)
+                    .map_err(|e| VersionError::Io(e, PathBuf::from(p))));
+                Ok(())
+            }));
+            hash.field("path", self.path.to_str().unwrap());
         } else {
             // We don't version the files outside of the /work because
             // we believe they are result of the commands run above
             //
             // And we need already built container to version the files
             // inside the container which is ugly
+            hash.field("source", src.to_str().unwrap());
+            hash.field("path", self.path.to_str().unwrap());
+            if !self.preserve_permissions {
```

Comment:
Should we hash `self.preserve_permissions` flag? If `preserve_permissions` is `true` we will hash `self.umask` so hashes will be different.